src/providers/IProvider.ts

```diff
@@ -8,61 +8,156 @@
-
+/**
+ * Provider Type to be extended for implmenting new providers
+ *
+ * @export
+ * @abstract
+ * @class IProvider
+ * @implements {AuthenticationProvider}
+ */
 export abstract class IProvider implements AuthenticationProvider {
+  // tslint:disable-next-line: completed-docs
+  public graph: Graph;
   private _state: ProviderState;
   private _loginChangedDispatcher = new EventDispatcher<LoginChangedEvent>();
-
+  /**
+   * returns state of Provider
+   *
+   * @readonly
+   * @type {ProviderState}
+   * @memberof IProvider
+   */
   public get state(): ProviderState {
     return this._state;
   }
 
+  constructor() {
+    this._state = ProviderState.Loading;
+  }
+
+  /**
+   * sets state of Provider and fires loginchangedDispatcher
+   *
+   * @param {ProviderState} state
+   * @memberof IProvider
+   */
   public setState(state: ProviderState) {
     if (state !== this._state) {
       this._state = state;
       this._loginChangedDispatcher.fire({});
     }
   }
 
-  // events
+  /**
+   * event handler when login changes
+   *
+   * @param {EventHandler<LoginChangedEvent>} eventHandler
+   * @memberof IProvider
+   */
   public onStateChanged(eventHandler: EventHandler<LoginChangedEvent>) {
     this._loginChangedDispatcher.add(eventHandler);
   }
-
+  /**
+   * removes event andler for when login changes
+   *
+   * @param {EventHandler<LoginChangedEvent>} eventHandler
+   * @memberof IProvider
+   */
   public removeStateChangedHandler(eventHandler: EventHandler<LoginChangedEvent>) {
     this._loginChangedDispatcher.remove(eventHandler);
   }
 
-  constructor() {
-    this._state = ProviderState.Loading;
-  }
-
+  /**
+   * login
+   *
+   * @returns {Promise<void>}
+   * @memberof IProvider
+   */
   public login?(): Promise<void>;
+
+  /**
+   * logout
+   *
+   * @returns {Promise<void>}
+   * @memberof IProvider
+   */
   public logout?(): Promise<void>;
+
+  /**
+   * uses scopes to recieve access token
+   *
+   * @param {...string[]} scopes
+   * @returns {Promise<string>}
+   * @memberof IProvider
+   */
   public getAccessTokenForScopes(...scopes: string[]): Promise<string> {
     return this.getAccessToken({ scopes });
   }
 
+  /**
+   * Promise to recieve access token using Provider options
+   *
+   * @abstract
+   * @param {AuthenticationProviderOptions} [options]
+   * @returns {Promise<string>}
+   * @memberof IProvider
+   */
   public abstract getAccessToken(options?: AuthenticationProviderOptions): Promise<string>;
-  public graph: Graph;
+  // tslint:disable-next-line: completed-docs
 }
 
+// tslint:disable-next-line: completed-docs
 export type EventHandler<E> = (event: E) => void;
+
+/**
+ * loginChangedEvent
+ *
+ * @export
+ * @interface LoginChangedEvent
+ */
+// tslint:disable-next-line: no-empty-interface
 export interface LoginChangedEvent {}
 
+/**
+ * Provider EventDispatcher
+ *
+ * @export
+ * @class EventDispatcher
+ * @template E
+ */
+// tslint:disable-next-line: max-classes-per-file
```

Comment:
We should split up classes to follow best practices whenever we can

Reply:
478fe6f